text/0000-deployment-language.md

```diff
@@ -0,0 +1,561 @@
+end;
+```
+
+
+#### Required Tremor Script Changes
+
+In order to make defining config entries in tremor-script convenient, we need to add a few features to tremor-script to make this work reasonably well in a configuration context, where all we want is to return a record value without too much fuss and ceremony:
+
+* Add multi-let statements, that combine multiple `let`s inside a single statement whcih defined the variables therein and returns a record value with all those definitions.
+* Add auto-creation of intermediary path segments in let stetements:
+
+  ```
+  let config.nested.value = 1;
+  ```
+
+  In this case if `config.nested` does not exist we would auto-create it as part of this let as an empty record.
+
+### Pipelines
+
+Pipelines are defined using a unique pipeline id, optionally some arguments and a mandatory query block that embeds a trickle query:
+
+EBNF Grammar:
+
+```
+pipeline      = "define" "pipeline" pipeline_id
+                [ "args" argument_list ]
+                "query" 
+                trickle_query
+                "end"
+pipeline_id   = id
+argument_list = ( id | id "=" expr ) [ "," argument_list ]
+```
+
+Whitespace and newlines are not significant here.
+
+**Example:**
+
+`args` can be referenced within the trickle query via the `args` root. As these are filled upon creation/instantiation
+a pipeline definition becomes a template:
+
+```
+define pipeline my_pipeline
+args
+    required_argument,
+    optional_arg_with_default = "default value"
+query
+    use std::datetime;
+    define tumbling window fifteen_secs
+    with
+        interval = datetime::with_seconds(args.required_argument),
+    end;
+
+    select { "count": aggr::stats::count(event) } from in[fifteen_secs] into out having event.count > 0;
+end; # we probably need to find a different way to terminate an embedded trickle query or find some trick
+```
+
+It might be interesting to be able to load a trickle query from a trickle file.
+To that end we add new config directives to trickle that can define arguments and their default values.
+
+```trickle
+#!config args my_arg = "foo", required_arg
+...
+```
+
+### Flow
+
+Flows are a new type of artefact that incorporates the previous concepts of `binding` and `mapping`.
+Flows define how artefacts are connected to form an event flow from sources via pipelines towards sinks.
+
+EBNF Grammar:
+
+```
+FLOW          = "define" "flow" flow_id
+                [ "args" argument_list ]
+                "links"
+                  link_list
+                "end"
+flow_id       = id
+argument_list = ( id | id "=" expr ) [ "," argument_list ]
+link_list     = link ";" [ link_list ]
+link          = "connect" tremor_url "to" tremor_url
+instance_port = artefact_id [ "/" instance_id ] ":" port
+```
+
+Example:
+
+```
+define flow my_eventflow
+args
+    required_arg,
+    optional_arg = "default value"
+links
+    connect "tremor://onramp/my_source/{required_arg}:out" to my_pipeline:in;
```

Comment:
It feels wrong to use `:` here for seperating entity and port and `/` in other places. It should be the same everywhere

Reply:
yeah, i initially thought about changing the tremor url syntax, but got away from it. bad matthias.

Reply:
It's worth noting that  `tremor://.../<id>:<port>` doesn't have a meaning for the URL it's simply stringly typed, `/` as we used before seems to be the better choice to me

Reply:
The reason for coming up with `:` in the first place was that we have ambiguities when parsing relative urls (without `tremor://`). It hard to know if a url has a port or not, in the parse function we take a guess. The `:` would resolve this. But nevermind, not in scope for this RFC.